### ISSUE TITLE
Cannot insert module with "insmod"

### ISSUE THREAD

**Issue by thalatta** [akiradeveloper/dm-writeboost#186]
I am trying to compile and insert the dm-writeboost kernel module without dkms. Using dkms to install and uninstall takes too long for developing purposes. For developing, I want to be able to use "sudo insmod dm-writeboost.ko".

Module works fine with dkms. After installing with dkms, "sudo modprobe dm-writeboost" creates entry in /proc/modules.

However, using "sudo insmod dm-writeboost.ko" results in error "insmod: ERROR: could not insert module dm-writeboost.ko: Unknown symbol in module."

To check the unknown symbol, I checked /var/log/syslog. It has this entry: "dm_writeboost: Unknown symbol crc32c (err 0)." This is in the Linux kernel tree. I can locate the crc32c.h include file in my Linux kernel tree, where crc32c is declared.

Remember, dkms with modprobe works correctly. So why does insmod command not work?

**Comment by akiradeveloper** [comment]
insmod loads the module only while modprobe loads dependencies too.

**Comment by thalatta** [comment]
Thanks for the reply. I will try to load the module with modprobe after reading "The Linux Kernel Module Programming Guide" further.

Instead of closing this issue, I would like to keep it open for more questions I might have. That way, I don't need to open a new issue every time I have a question.

Thanks!

**Comment by thalatta** [comment]
I've tried searching with grep and other ways to find where the struct dm_target is defined, but all that can be found on my part is a declaration in include/linux/device-mapper.h.

Can I get a hint as to where dm_target is defined?

**Comment by thalatta** [comment]
I created a question on stackoverflow asking about a piece of code I didn't understand.

https://stackoverflow.com/questions/46880585/what-does-the-is-aligned-macro-in-the-linux-kernel-do

What does the IS_ALIGNED macro for? I know what it does. It is defined in the linux kernel source as
`#define IS_ALIGNED(x, a)        (((x) & ((typeof(x))(a) - 1)) == 0)`

It is used in dm-writeboost as 
`IS_ALIGNED(addr, 1 << 9)`

This left-shifts 1 by 9, making 11111111. Then it does bitwise-and with x. How does this test for alignment?

**Comment by akiradeveloper** [comment]
@thalatta Good question. I read the stackoverflow quickly and it seems the answer is correct. The code was introduced back before v1.0.0 when dm-writeboost was quite experimental.

Writeboost have ever used vmalloc to allocate buffer and I found it it sometimes return unaligned buffer. I removed the vmallocs and replaced them by kmem_pools that should always align. The check is to make sure the allocated buffers are always aligned.

**Comment by thalatta** [comment]
Thank you for elaborating

**Comment by thalatta** [comment]
I am trying to understand what dm_per_bio_data does. Link to Stackoverflow page I created here: https://stackoverflow.com/questions/46919206/what-does-the-kernel-function-dm-per-bio-data-do

In device-mapper.h, we have the following code:
`struct dm_target_io {
    struct dm_io *io;
    struct dm_target *ti;
    unsigned target_bio_nr;
    unsigned *len_ptr;
    struct bio clone;
};

static inline void *dm_per_bio_data(struct bio *bio, size_t data_size)
{
    return (char *)bio - offsetof(struct dm_target_io, clone) - data_size;
}`

So dm_per_bio_data begins with the location of bio, then subtracts offset to get location of dm_target_io. Then it subtracts data_size. Where does the pointer go now?

**Comment by akiradeveloper** [comment]
replied directly because it's not a Writeboost specific question. If you truly want to know about device-mapper. Please read the code carefully and widely. device-mapper doesn't provide enough documentation because they assumes target implementers are capable of doing without it.

**Comment by thalatta** [comment]
Thank you. I would not have found the code without your help

**Comment by allenjhan** [comment]
I have a question about wb->nr_dirty_caches. This variable is atomically initialized, incremented, and decremented. Is this because of possible interrupt?

**Comment by akiradeveloper** [comment]
No. For the concurrency control. The variable is changed by multiple threads. It's incremented in the main thread and decremented in the background thread, `mark_clean_seg` for example.

**Comment by allenjhan** [comment]
Okay, thanks

**Comment by allenjhan** [comment]
The abbreviation "nr" appears often in the codebase. What does it mean?

**Comment by akiradeveloper** [comment]
number

**Comment by allenjhan** [comment]
Thanks

**Comment by allenjhan** [comment]
Sorry, I have a question about the code.

In "init_read_cache_cells," we have
`INIT_WORK(&wb->read_cache_work, read_cache_proc);`

Why does "read_cache_proc" require its own thread?

**Comment by allenjhan** [comment]
I looked up what the INIT_WORK() macro does and that cleared up my confusion. I thought INIT_WORK() was starting the separate thread, but actually, it only initializes the work_struct structure.

**Comment by allenjhan** [comment]
I am trying to understand the dm_io() interface. I found this blog entry, but I cannot read Japanese and Google Translate only helps a little: http://akiradeveloper.hatenadiary.com/entry/2013/05/18/132406

Mainly, I want to understand the code in read_backing_async().

	io_req = (struct dm_io_request) {
		WB_IO_READ,
		.client = wb->io_client,
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,14,0)
		.mem.type = DM_IO_BIO,
		.mem.ptr.bio = bio,
#else
		.mem.type = DM_IO_BVEC,
		.mem.ptr.bvec = bio->bi_io_vec + bio->bi_idx,
#endif
		.notify.fn = read_backing_async_callback,
		.notify.context = ctx
	};
	region = (struct dm_io_region) {
		.bdev = wb->backing_dev->bdev,
		.sector = bi_sector(bio),
		.count = 8
	};

Am I understanding this correctly? This code will copy the code from backing_dev's bi_sector(bio) to bio. And .count is the number of sectors to read.

**Comment by akiradeveloper** [comment]
@allenjhan Sorry for inconvenience. The post is about what is happening under the hood of `dm_io` abstraction.

Your understanding is correct. The code is to achieve the followings:

1. read the data from backing store
2. copy the data to read cache cell

The codes in read-caching is very complicated. If you find a bug, please report it.

**Comment by allenjhan** [comment]
Thank you. :)

**Comment by allenjhan** [comment]
In  dm_writeboost_target.c, there is the function `reserve_read_cache_cell`. It has this code:

    if (!bio_is_fullsize(bio))
        return false;

I understand that we are discarding partial data smaller than 4KB here, but what if bio is larger than 4KB? Is that possible, and would the data be discarded too if bio is larger than 4KB?

**Comment by akiradeveloper** [comment]
@allenjhan No. Writeboost splits all in-coming bio into smaller bios no bigger than 4KB. Most of them, especially if limited to read requests, are 4KB because of the page cache. By removing the chance of caching even smaller fragments, the code becomes really simpler and faster.
 
This is realized by using the device-mapper's functionality which is set here 

```c
static int init_core_struct(struct dm_target *ti)
{
	int err = 0;
	struct wb_device *wb;

	err = dm_set_target_max_io_len(ti, 1 << 3);
```

**Comment by allenjhan** [comment]
I want to ask a follow-up question about 4KB data.

In the code for `initialize_write_io()`, will `bio_sectors(bio)` always be 8?

    static void initialize_write_io(struct write_io *wio, struct bio *bio)
    {
        u8 offset = bio_calc_offset(bio);
        sector_t count = bio_sectors(bio);
        copy_bio_payload(wio->data + (offset << 9), bio);
        wio->data_bits = to_mask(offset, count);
    }

If `bio_sectors(bio)` is always 8, then that means that we are reading across the 4KB boundary if offset is not zero.

**Comment by akiradeveloper** [comment]
@allenjhan No.

**Comment by allenjhan** [comment]
Sorry, my mistake. The fixed 4KB size is for the read cache only. Write cache does not have this restriction.

There is something I don't understand about the bio struct. I spent some time reading chapter 14 of _Linux Kernel Development_ by Robert Love. It covers the bio struct, but does not answer my question. Why does the bio_struct have a field called bi_size, if the bio_vec array specifies a bv_len value for each bio_vec? How is the bv_len value of the bio_vec different from the bi_size of the whole bio?

**Comment by akiradeveloper** [comment]
@allenjhan I think this is not the right place to ask the question because it's too general.

**Comment by allenjhan** [comment]
Okay, I created a Stackoverflow page. The link is here: https://stackoverflow.com/questions/48208991/how-are-bv-len-and-bi-size-each-used-in-struct-bio-of-the-linux-kernel

**Comment by allenjhan** [comment]
I found the answer in `copy_bio_payload()`. The sum total of all the `bv_len` fields in their respective `bio_vec` will equal the value of the `bi_size` field in `bio`.

**Comment by allenjhan** [comment]
I have another question that is a little general, so I created a Stackoverflow page for it. Link is here: https://stackoverflow.com/questions/48290260/what-is-the-difference-in-linux-kernel-device-mapper-between-dm-mapio-submitte

It concerns difference between DM_MAPIO_SUBMITTED and DM_MAPIO_REMAPPED.

**Comment by allenjhan** [comment]
By examining the bio_remap() function, I think I understand DM_MAPIO_REMAPPED now.

**Comment by akiradeveloper** [comment]
@allenjhan  About the flags, I recommend you to read the kernel code under driver/md. It was much unreadable code when I started this project but it has been refactored and bit readable now. device-mapper lacks documentation and target implementors need to read the code but they are better than other codebase in Linux kernel.
